chips/earlgrey/src/chip.rs

```diff
@@ -110,6 +114,25 @@ impl<'a, A: 'static + Alarm<'static>, I: InterruptService<()> + 'a> EarlGrey<'a,
 
     unsafe fn handle_plic_interrupts(&self) {
         while let Some(interrupt) = self.plic.get_saved_interrupts() {
+            // Claim the interrupt before we handle it.
+            // Currently the interrupt has been claimed but not completed.
+            // This means that if the interrupt re-asserts we will loose the
+            // re-assertion. Generally this isn't a problem, but some of the
+            // interrupt handlers expect that interrupts could occur.
+            // For example the HMAC interrupt handler will write data to the
+            // HMAC buffer. We then rely on an interrupt triggering when that
+            // buffer becomes empty. This can happen while we are still in the
+            // interrupt handler. To ensure we don't loose the interrupt we
+            // claim it here.
+            // In order to stop an interrupt loop, we first disable the
+            // interrupt. `service_pending_interrupts()` will re-enable
+            // interrupts once they are all handled.
+            self.atomic(|| {
```

Comment:
I am a little worried about this logic.

In many cases, it seems like it's OK for the interrupt to be completed after the full handler. In the HMAC case, it's important that the interrupt be completed earlier. But I think it is easy to come up with cases in which you don't want it completed earlier, e.g., a level interrupt on GPIO.

Since we probably don't want to go redesigning abstractions for this, I'd suggest just special casing this for now. That is, just special case on HMAC here and leave the logic for other interrupts unchanged.